lib/ansible/modules/network/panos/panos_cert_gen_ssh.py

```diff
@@ -0,0 +1,200 @@
+            - number of bits used by the RSA alg
+        required: false
+        default: "1024"
+'''
+
+EXAMPLES = '''
+# Generates a new self-signed certificate using ssh
+- name: generate self signed certificate
+  panos_cert_gen_ssh:
+    ip_address: "192.168.1.1"
+    password: "paloalto"
+    cert_cn: "1.1.1.1"
+    cert_friendly_name: "test123"
+    signed_by: "root-ca"
+'''
+
+RETURN='''
+# Default return values
+'''
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'version': '1.0'}
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.basic import get_exception
+import time
+
+try:
+    import paramiko
+    HAS_LIB=True
+except ImportError:
+    HAS_LIB=False
+
+_PROMPTBUFF = 4096
+
+
+def wait_with_timeout(module, shell, prompt, timeout=60):
+    now = time.time()
+    result = ""
+    while True:
+        if shell.recv_ready():
+            result += shell.recv(_PROMPTBUFF)
+            endresult = result.strip()
+            if len(endresult) != 0 and endresult[-1] == prompt:
+                break
+
+        if time.time()-now > timeout:
+            module.fail_json(msg="Timeout waiting for prompt")
+
+    return result
+
+
+def generate_cert(module, ip_address, key_filename, password,
+                  cert_cn, cert_friendly_name, signed_by, rsa_nbits ):
+    stdout = ""
+
+    client = paramiko.SSHClient()
+
+    # add policy to accept all host keys, I haven't found
+    # a way to retreive the instance SSH key fingerprint from AWS
+    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+
+    if not key_filename:
+        client.connect(ip_address, username="admin", password=password)
+    else:
+        client.connect(ip_address, username="admin", key_filename=key_filename)
+
+    shell = client.invoke_shell()
+    # wait for the shell to start
+    buff = wait_with_timeout(module, shell, ">")
+    stdout += buff
+
+    # generate self-signed certificate
+    if isinstance(cert_cn, list):
+        cert_cn = cert_cn[0]
+    cmd = 'request certificate generate signed-by {0} certificate-name {1} name {2} algorithm RSA rsa-nbits {3}\n'.format(
+        signed_by, cert_friendly_name, cert_cn, rsa_nbits)
+    shell.send(cmd)
+
+    # wait for the shell to complete
+    buff = wait_with_timeout(module, shell, ">")
+    stdout += buff
+
+    # exit
+    shell.send('exit\n')
+
+    if 'Success' not in buff:
+        module.fail_json(msg="Error generating self signed certificate: "+stdout)
+
+    client.close()
+    return stdout
+
+
+def main():
+    argument_spec = dict(
+        ip_address=dict(required=True),
+        key_filename=dict(),
+        password=dict(),
+        cert_cn=dict(required=True),
+        cert_friendly_name=dict(required=True),
+        rsa_nbits=dict(default='2048'),
+        signed_by=dict(required=True)
+
+    )
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=False,
```

Comment:
I've been thinking about that and what we need to do is implement a functionality that will un-roll the last transaction on the device. If we do this we have to do it across all modules. This is something I plan to do for V2.0.